### ISSUE TITLE
Filtered Stream Stop Causing Exceptions

### ISSUE THREAD

**Issue by PierreJC** [linvi/tweetinvi#149]
Hello,

I am facing an issue with filtered streams within my C# program. I am trying to start a stream for a specific track, and then whenever the features are met, I call stop on the stream, in order to reset the stream for another track. The weird thing is that every single time I call stop on the stream, I get multiple IO first chance exceptions and a socket exception.

Also I try to override the functionality sometimes to just reset the stream for another track, by stopping the current stream and regenerating a new one, and then it would cause all these exceptions plus another web exception that stops the next stream from working.

I am StartStreamMatchingAllConditionsAsync everytime I start it again.
Would you be able to know what's happening here ?

Thank you.

**Comment by linvi** [comment]
Hello,

First of all could you let me know which version of Tweetinvi you are using as I just released a new version (0.9.11.0) today.

Secondly you should never have a stream throwing an error in the main application (thread calling `StartStreamMatchingAllConditionsAsync`).

When an error is received in a `FilteredStream`, the `StreamStopped` event will be raised and the error will be in the even args ([Stream Error Handling](https://github.com/linvi/tweetinvi/wiki/Streams#stop-and-error-handling)).

Feel free to share your code so that I can help you more.

Thanks,
Linvi

**Comment by PierreChahwan** [comment]
Hi again,

I just updated to the latest version. Now the thing is those first chance exceptions are only occurring in the output window which shouldn't be that big of a deal. But the big issue is when I try to override in order to reset the stream to a different track, it gives me the problem of a web exception in the output window and stops grabbing tweets.

Below is the code of what I am doing, this gets fired when I click an override button to reset and start a new stream : 

`   private void overrideBT_Click(object sender, EventArgs e)
        {
               // Stop the old stream
               m_ifsCurrentStream.StopStream();

                // Create new stream
                m_ifsCurrentStream = Tweetinvi.Stream.CreateFilteredStream(m_tcTwitterCredentials);

               // Set new track
               m_ifsCurrentStream.AddTrack(m_hashTag);

              // Set event handler
                m_ifsCurrentStream.MatchingTweetReceived += _eventHandler;

              // Run the new stream
               m_ifsCurrentStream.StartStreamMatchingAllConditionsAsync();
        }
`

**Comment by PierreChahwan** [comment]
Just tried to catch the exception, it says that I "exceeded the connection limit per user". Now when I do it the normal way, where I send tweets to trigger the change, it works fine. But my override button clicks seem to cause this. Any advice on how to solve this one ?

**Comment by linvi** [comment]
"exceeded the connection limit per user" means that you are running 2 `FilteredStream` at the same time. Which is not allowed by Twitter.

Instead of instantiating the stream every time could you try to :

``` c#

constructor() 
{
              // Create new stream done in the constructor
              m_ifsCurrentStream = Tweetinvi.Stream.CreateFilteredStream(m_tcTwitterCredentials);

              // Set event handler
              m_ifsCurrentStream.MatchingTweetReceived += _eventHandler;
}

private void overrideBT_Click(object sender, EventArgs e)
{
               // Stop the old stream
               m_ifsCurrentStream.StopStream();

               // Set new track
               m_ifsCurrentStream.AddTrack(m_hashTag);

              // Run the new stream
               m_ifsCurrentStream.StartStreamMatchingAllConditionsAsync();
}
```

**Comment by PierreJC** [comment]
Just tried that and it still does it. 
Do you think it might be that streams are still running in the background ? Even though I'm doing everything we discussed above ?

**Comment by linvi** [comment]
I have been able to reproduce the problem I will work out for a fix.

**Comment by linvi** [comment]
Could you let me know if this work for you?

[Tweetinvi 0.9.11.1 - Binaries.zip](https://github.com/linvi/tweetinvi/files/160168/Tweetinvi.0.9.11.1.-.Binaries.zip)
[Tweetinvi 0.9.11.1 - Merged Binaries.zip](https://github.com/linvi/tweetinvi/files/160169/Tweetinvi.0.9.11.1.-.Merged.Binaries.zip)

**Comment by linvi** [comment]
The code does not solve the problem. I will get back to you when I have a working solution for the problem.

Please use version 0.9.10.2 temporarily.

**Comment by PierreJC** [comment]
Hello again,

Alright thank you very much for taking the time to look into the issue. I'll be waiting for your response. Meanwhile, i'll work on some other parts of the program.

**Comment by PierreJC** [comment]
Hey linvi,

I might have found something that can help fix this issue. The thing is if we call StopStream(), and then right after StartStreamMatchingAllConditions(), it throws the web exception.
However, if I call stop stream on one iteration, and then after 1 or 2 iterations, call start stream, it wont cause this issue to happen.

My guess would be maybe when you stop the stream you are triggering a bool that's being used in the asynchronous function, while this function is always running in the background, and the bool is not getting enough time to make the async function break out.

Let me know if this can be of any help.

**Comment by PierreJC** [comment]
Nevermind, still does it. Didn't do it on the first run, now it does. Weird.

**Comment by linvi** [comment]
Hi,

Thank you for this. I did noticed that already. I think it is related with the `HttpClient.SendAsync` to execute some of its long running operation in the `ThreadPool` that is being filled.

I currently cannot work much on `Tweetinvi` but I will do my very best to fix this issue as soon as I can.

If you want to try and find where the issue is located you can download the project and in your main project add references to all the Tweetinvi.* projects.

To reproduce the problem create a where clause stopping and restarting a same stream.
After 6 calls, it will throw an exception line 218 of `StreamTask.cs` :

``` c#
await client.SendAsync(request, HttpCompletionOption.ResponseHeadersRead).ConfigureAwait(false);
```

Which I think is coming from the ThreadPool.

**Comment by PierreJC** [comment]
Hello again,

Alright, I will try my best to try and debug the issue, and if I get to a solution before you do, I will let you know as soon as possible. It would also be awesome if you can also let me know if you find one.

Thank you.

**Comment by linvi** [comment]
No worry, this is my "job" to fix the issues in the library. I spent 2h30 tonight but no successful result for a potential fix.

What I have been able to notice up to now is that when you call `StopStream` the problem occurs.

First fix that help resolving the Json issue is the following : 

``` c#
// Remove the following line 353 of StreamTask.cs
 _currentStreamReader.Dispose();
```

Now what is happening, if you open your thread pane (Debug -> Window -> Threads).
What you will notice is that the application will work fine up to the point when a Thread exits.

Here is the easier repro code:

``` c#
var stream = Stream.CreateSampleStream();

stream.TweetReceived += (sender, args) =>
{
    Console.WriteLine(args.Tweet);
};

while (true)
{
    stream.StartStream();

    Thread.Sleep(2000);

    Console.Clear();

    stream.StopStream(); // At the 6th iteration the .SendAsync will throw TaskCancelledException
}
```

**Comment by PierreJC** [comment]
Hey linvi,

Just wanted to tell you that after some debugging and research, it seems like this is normal behavior in our case. I tried using other libraries to replicate the issue and it still happened (Linq2Twitter for example). The issue is that those fast stops and restarts get rejected by Twitter after a couple of times. How I fixed it now is a delay of 15s between a stop and restart because it suits the purpose of my program. You might need less of a delay, but this has fixed my problem.

```
m_twitterInterface.StopStream();
m_resetTimer.Stop();

private void restartTimerEventHandler(object sender, ElapsedEventArgs e)
{
     m_resetTime.Start();
}

// This reset timer of course goes for 15 seconds, and on elapsed, it starts the stream again.
```

**Comment by linvi** [comment]
Thank you for this, but I will let this problem open as I think it is a Task issue with the Microsoft.Bcl.Async mechanism. I am currently working on a repro project.

Depending on the complexity it might be fixed in 0.9.12 instead of 0.9.11.1.

**Comment by PierreJC** [comment]
No problem ! That might be an issue too. Also need to add something that might help you. When i added 15 seconds to fix the issue, in those 15 seconds, you can see threads exiting on the output window before i restart the stream. However, if I dont wait, it doesnt always exit as fast. So it might be something with threads not exiting at the correct time or like you said related to the Task issue.

**Comment by linvi** [comment]
I am still working on the problem. I created a repository which I hope will help me fix this problem.

You can find it [here](https://github.com/linvi/ConcurrentStreamTests) if you are interested.

**Comment by linvi** [comment]
Moved to next release as current is getting late.

**Comment by linvi** [comment]
Closed for now as this issue is very old. Feel free to reopen it if this has not been solved by the latest release of Tweetinvi (2.0).

**Comment by linvi** [comment]
I have tested the concurrent stream issue with .NET 4.6 and .NET Standard 1.3.
The bug is no longer there. Therefore 2.0 has fixed this problem.
